libcnb/src/layer/public_interface.rs

```diff
@@ -0,0 +1,187 @@
+
+    /// This method will be called by libcnb when the layer already exists to determine the strategy
+    /// used to deal with it. Implementations of this method can use the current layer contents and
+    /// metadata to make that decision.
+    ///
+    /// This can be used to invalidate a layer based on metadata. For example, the layer metadata
+    /// could contain a language runtime version string. If the version requested by the user is
+    /// different, this method should can return [`ExistingLayerStrategy::Recreate`], causing a new
+    /// language runtime version to be installed from scratch.
+    ///
+    /// It can also be used cause a call to [`update`](Layer::update), updating the contents of the
+    /// existing layer. Installing additional application dependencies with a package manager is
+    /// one common case where this strategy makes sense. Implementations need to return
+    /// [`ExistingLayerStrategy::Update`] to trigger that behavior.
+    ///
+    /// When not implemented, the default implementation will return
+    /// [`ExistingLayerStrategy::Keep`], causing the layer to stay as-is (i.e. no calls to either
+    /// [`create`](Layer::create) nor [`update`](Layer::update)).
+    ///
+    /// # Implementation Requirements
+    /// Implementations **MUST NOT** modify the file-system.
+    fn existing_layer_strategy(
+        &self,
+        context: &BuildContext<Self::Buildpack>,
+        layer_data: &LayerData<Self::Metadata>,
+    ) -> Result<ExistingLayerStrategy, <Self::Buildpack as Buildpack>::Error> {
+        Ok(ExistingLayerStrategy::Keep)
+    }
+
+    /// Updates the layer contents and metadata based on the cached version of a previous run.
+    ///
+    /// `layer_path` will be a directory with the data from a previous run. This method can modify
+    /// the contents freely. Layer metadata, including environment variables, is part of the return
+    /// value of this function and will be written to the appropriate locations by libcnb
+    /// automatically.
+    ///
+    /// The return value of this method is the canonical value for metadata and environment variables.
+    /// If the returned [`LayerResult`](LayerResult) does not contain metadata or environment
+    /// variables, the resulting layer will not have either. To keep the values from the cached
+    /// version you need to explicitly add them to the result. This can be done by reading that
+    /// data from the given [`LayerData`](LayerData) value.
+    ///
+    /// # Implementation Requirements
+    /// Implementations **MUST NOT** write to any other location than `layer_path`.
+    fn update(
+        &self,
+        context: &BuildContext<Self::Buildpack>,
+        layer_data: &LayerData<Self::Metadata>,
+    ) -> Result<LayerResult<Self::Metadata>, <Self::Buildpack as Buildpack>::Error> {
+        LayerResultBuilder::new(layer_data.content_metadata.metadata.clone()).build()
```

Comment:
It seems that either:
1. `update()` never gets called (because `cache=false` or `existing_layer_strategy` returns something other than `ExistingLayerStrategy::Update`.
2. `update()` *will* be called, but the default implementation will never be what was intended (otherwise why not use `ExistingLayerStrategy::Keep`?), and as such not implementing `update()` in the buildpack is a bug.

Given the above, should the default implementation for `update()` be `unimplemented!("...")`, so as to make this bug more obvious to buildpack authors?

Reply:
@schneems pointed out that having the default implementation of `update` being `unimplemented!` might cause the buildpack to panic in a subsequent run and the buildpack author might not be aware. Worst case scenario, they would just test it once, see that it works and ship it. All in all, should we want to prevent this, we should prevent compilation instead of failing at runtime.

Reply:
Looking at this from a different angle: Currently, if they don't implement `update`, the update process is a no-op. That is IMO reasonable and expected. If I, for example, want to log something I have to call my `log()` function. And if I don't it does not log. I acknowledge the case is slightly different since they expressed the intent to update via `Layer::existing_layer_strategy`. But that method does not return any logic and I don't think anyone would expect that the layer is updated in a way they want without writing the code for it.

Reply:
@Malax Won't it drop their env vars at present, meaning their buildpack could fail at runtime anyway?

Reply:
So perhaps we should be using `unimplemented!` here, then have the more opinionated Traits as the recommended approach, and using `Layer` is left for advanced use-cases where `unimplemented!` is fine?

Reply:
I'm very opposed to use `unimplemented!` in library code. If we want to force implementations we should not provide a default one and make the compilation fail instead.

Reply:
> Won't it drop their env vars at present?

Yes, and that's a bug.

Reply:
> we should not provide a default one and make the compilation fail instead.

Which I guess we can do once we have the more opinionated traits - so yeah I agree.

Reply:
Default `update` bug WRT env vars fixed in 9dbf6e3b1f1bf38b8268c1acfba9ed61ab7754d2 (No tests, will be added in #169)

Reply:
@Malax So what made me think the choice to not return `env` was intentional, was this comment:

```rust
    /// The return value of this method is the canonical value for metadata and environment variables.
    /// If the returned [`LayerResult`](LayerResult) does not contain metadata or environment
    /// variables, the resulting layer will not have either. To keep the values from the cached
    /// version you need to explicitly add them to the result. This can be done by reading that
    /// data from the given [`LayerData`](LayerData) value.
```

Specifically the "To keep the values from the cached version you need to explicitly add them to the result".

It seems the comment needs changing to match the implementation?

Reply:
Oh, I guess the docs are saying "the function has to explicitly add them, but we do that for you in the default implementation". Not "if you want them, you'll have to explicitly add them yourself".

Reply:
So perhaps `you need to explicitly add them` -> `they must be explicitly added to the result, as seen in the default implementation`.

(This is a classic case where "you" having both singular and plural meanings can create ambiguity and other languages don't have this problem :-))

Reply:
Does it? The comment is intended to say that implementations of `update` (regardless of default or custom) need to explicitly return the env they want this layer to have after the update. 

The default implementation is intended to be a no-op, so it has to keep the env from the previous run. That's why we now add it to the result.

Reply:
I think our comments raced :-)

To add some extra detail - in the current wording I interpreted "you" to mean "you singular" as in the reader of the docs, ie: the buildpack author -- rather than "we" or "one" etc.

Reply:
> I think our comments raced :-)

Yeah, they did race :D 

I made some changes to the docs based on your suggestion in 25125a9e17d0c9ec3c8fa511eff1090e614de589.

Reply:
Yeah those changes are great, thank you :-)